shacl12-core/index.html

```diff
@@ -1587,267 +1585,230 @@ <h3>Property Shapes</h3>
+						SPARQL <code>ZeroOrOnePath</code> with <code>path(v,G)</code> as its <code>elt</code>.
+					</p>
+				</section>
+			</section>
 
-ex:AddressGroup
-	a sh:PropertyGroup ;
-	sh:order 1 ;
-	rdfs:label "Address" .
-								</div>
-							</div>
-						</aside>
-						<p>
-							A form building application MAY use the information above to display information as follows:
-						</p>
-						<div style="background: #f3f3f3; padding: 8px">
-							<div style="font-size: 18px; color: #0000a0"><b>Name</b></div>
-							<table>
-								<tr>
-									<td style="text-align: right; width: 160px"><b>first name:</b></td>
-									<td>John</td>
-								</tr>
-								<tr>
-									<td style="text-align: right"><b>last name:</b></td>
-									<td>Doe</td>
-								</tr>
-							</table>
-							<div style="font-size: 18px; padding-top: 6px; color: #0000a0"><b>Address</b></div>
-							<table>
-								<tr>
-									<td style="text-align: right; width: 160px"><b>street address:</b></td>
-									<td>123 Silverado Ave</td>
-								</tr>
-								<tr>
-									<td style="text-align: right"><b>locality:</b></td>
-									<td>Cupertino</td>
-								</tr>
-								<tr>
-									<td style="text-align: right"><b>zip code:</b></td>
-									<td>54321</td>
-								</tr>
-							</table>
+			<section id="node-expressions">
+				<h3>Node Expressions</h3>
+				<p>
+					This section introduces the concept of <a>node expressions</a> as an extension point used by other specifications such as <a href="shacl12-sparql">SHACL-SPARQL</a>.
+					This SHACL Core specification does not define any <a>node expression types</a> but defines how they are generally handled.
+					<span class="todo">TODO: Add link to shacl12-node-expr once that is stable.</span>
+				</p>
+				<div class="def">
+					<div class="def-header">TEXTUAL DEFINITION</div>
+					A <dfn>node expression</dfn> is a function with exactly one <dfn>node expression type</dfn> as its type.
+					Each <a>node expression type</a> has an <a>IRI</a>.
+					Node expression types can declares one or more <dfn>node expression parameters</dfn>.
+					Each of these parameters has an <a>IRI</a>.
+					One of these parameters can be the <dfn>key parameter</dfn> that uniquely identifies the node expression type.
+					<br />
+					<br />
+					During <dfn>evaluation</dfn> a node expression takes the following input:
+					<ul>
+						<li>A <a>graph</a>, called the <dfn>active graph</dfn>.</li>
+						<li>An optional <a>node</a> called the <dfn>input focus node</dfn>.</li>
+						<li>One <a>node</a> as value for each of the declared parameters, optional for any but the key parameter unless stated otherwise in the definition of the node expression type.</li>
+					</ul>
+					The result of the evaluation of a node expression is a list of <a>nodes</a> (possibly empty and with duplicates) called the <dfn>output nodes</dfn>.
```

Comment:
_List_ of nodes, or _set_ of nodes?  The "possibly ... with duplicates" point would be moot for a set.  If a list, would there be conditions under which an order of returned nodes could be guaranteed?

Reply:
Yes it should be a list because then people can do sh:orderBy and sh:limit, e.g. when produced by sh:select

Reply:
`sh:order By` involves coordination between two lists. That is going to need some work. It is pushing the boundaries of what a node expression is.

```ttl
[
    sh:nodes [ sh:path ex:child ] ;
    sh:orderBy [ sh:path rdfs:label ] ;
] .
```
`sh:nodes` is a list of `?child`, not a list of pairs `(?child , ?label)`

`[ sh:path rdfs:label ]` is another list that might be a different length and/or may have matched so it does not arrive in the same sequence as `[ sh:path ex:child ]` so they can't be lined up to produce pairs.

```
[ sh:orderBy [ sh:path ex:child ] ]
```
may be more appropriate for NE without SPARQL.

Reply:
Yes I am not particularly fond of the old definition of sh:orderBy.